components/base/kinematicbase/kinematics.go

```diff
@@ -102,56 +107,106 @@ func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired
 	if err != nil {
 		return err
 	}
-	validRegion, err := ddk.newValidRegionCapsule(current, desired)
+	distErr, headingErr, err := ddk.errorState(current, desired)
+	if err != nil {
+		return err
+	}
+	validRegion, err := ddk.newValidRegionCapsule(current, desired, distErr)
 	if err != nil {
 		return err
 	}
 
+	t := time.NewTimer(timeout)
+	poseChange := make(chan bool)
+	movementErr := make(chan error)
+
 	// this loop polls the error state and issues a corresponding command to move the base to the objective
 	// when the base is within the positional threshold of the goal, exit the loop
-	for err = ctx.Err(); err == nil; err = ctx.Err() {
-		col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
-		if err != nil {
-			return err
-		}
-		if !col {
-			return errors.New("base has deviated too far from path")
-		}
+	utils.PanicCapturingGo(func() {
+		prevDistErr := -1
+		prevHeadingErr := math.Inf(-1)
+		var contextErr error
+		for contextErr = ctx.Err(); contextErr == nil; contextErr = ctx.Err() {
+			if poseChanged(prevDistErr, prevHeadingErr, distErr, headingErr) {
+				poseChange <- true
+			}
 
-		// get to the x, y location first - note that from the base's perspective +y is forward
-		desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
-		commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
-		if err != nil {
-			return err
-		}
+			col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
+			if err != nil {
+				movementErr <- err
+				return
+			}
+			if !col {
+				movementErr <- errors.New("base has deviated too far from path")
+				return
+			}
+
+			// get to the x, y location first - note that from the base's perspective +y is forward
+			desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
+			prevDistErr = distErr
```

Comment:
This seems like this can cause problems.

If you're iterating this loop rapidly, you may be sampling faster than the base can actually travel `distEpsilon` mm, in which case `poseChanged` will always be false.

I think you only want to update prevDistErr and prevHeadingErr if `poseChanged` returned true